File: vitral/src/vsdk/toolkit/environment/geometry/polyhedralBoundedSolidNodes/_PolyhedralBoundedSolidEdge.java
//===========================================================================
//=-------------------------------------------------------------------------=
//= Module history:                                                         =
//= - November 18 2006 - Oscar Chavarro: Original base version              =
//= - January 3 2007 - Oscar Chavarro: First phase implementation           =
//=-------------------------------------------------------------------------=
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =
//===========================================================================

package vsdk.toolkit.environment.geometry.polyhedralBoundedSolidNodes;

import vsdk.toolkit.common.Entity;
import vsdk.toolkit.environment.geometry.PolyhedralBoundedSolid;

/**
As noted in [MANT1988].10.2.2, a `_PolyhedralBoundedSolidEdge` makes a
face-to-face relationship representing the identification of the line
segments between faces.

Note that in the sake of simplify and eficiency current programming 
implementation this class exhibit public access attributes. It is important
to note that those attributes will only be accessed directly from related 
classes in the same package
(vsdk.toolkit.environment.geometry.polyhedralBoundedSolidNodes) and
from methods in the `PolyhedralBoundedSolid` class, and that they should
not be used from outer classes.
*/
public class _PolyhedralBoundedSolidEdge extends Entity {
    /// Check the general attribute description in superclass Entity.
    public static final long serialVersionUID = 20061118L;

    /// Reference to `right` half edge, as defined in [MANT1988].10.2.2.
    /// Note that half edge in this side is considered positively oriented.
    public _PolyhedralBoundedSolidHalfEdge rightHalf;

    /// Reference to `right` half edge, as defined in [MANT1988].10.2.2.
    /// Note that half edge in this side is considered negatively oriented.
    public _PolyhedralBoundedSolidHalfEdge leftHalf;

    //=================================================================
    public _PolyhedralBoundedSolidEdge(PolyhedralBoundedSolid parentSolid)
    {
        parentSolid.edgesList.add(this);
    rightHalf = null;
    leftHalf = null;
    }

    public String toString()
    {
        String msg;
        msg = "Edge, half1: ";
        if ( leftHalf == null ) {
        msg = msg + "null. ";
    }
    else {
            msg = msg + "vertex " + leftHalf.startingVertex.id;
    }

        msg = msg + " / half2: ";

        if ( rightHalf == null ) {
        msg = msg + "null. ";
    }
    else {
            msg = msg + "vertex " + rightHalf.startingVertex.id;
    }

        return msg;
    }
}

//===========================================================================
//= EOF                                                                     =
//===========================================================================

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/nodes/_PolyhedralBoundedSolidEdge.ts
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =

import { FundamentalEntity } from "../../../../../common/FundamentalEntity.js";
import { ColorRgb } from "../../../../../common/color/ColorRgb.js";
import type { PolyhedralBoundedSolid } from "../PolyhedralBoundedSolid.js";
import type { _PolyhedralBoundedSolidHalfEdge } from "./_PolyhedralBoundedSolidHalfEdge.js";

/**
As noted in [MANT1988].10.2.2, a `_PolyhedralBoundedSolidEdge` makes a
face-to-face relationship representing the identification of the line
segments between faces.

Note that in the sake of simplify and efficiency current programming
implementation this class exhibit public access attributes. It is important
to note that those attributes will only be accessed directly from related
classes in the same package
(vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.nodes) and
from methods in the `PolyhedralBoundedSolid` class, and that they should
not be used from outer classes.
*/
export class _PolyhedralBoundedSolidEdge extends FundamentalEntity {
    /// Reference to `right` half edge, as defined in [MANT1988].10.2.2.
    /// Note that half edge in this side is considered positively oriented.
    public rightHalf!: _PolyhedralBoundedSolidHalfEdge | null;

    /// Reference to `right` half edge, as defined in [MANT1988].10.2.2.
    /// Note that half edge in this side is considered negatively oriented.
    public leftHalf!: _PolyhedralBoundedSolidHalfEdge | null;

    //
    public id!: number;
    public debugColor!: ColorRgb;
    private static currentId = 1;

    //=================================================================
    public constructor(parentSolid: PolyhedralBoundedSolid) {
        super();
        this.init(parentSolid);
    }

    private init(parentSolid: PolyhedralBoundedSolid): void {
        parentSolid.getEdgesList().add(this);
        this.rightHalf = null;
        this.leftHalf = null;

        this.id = _PolyhedralBoundedSolidEdge.currentId;
        _PolyhedralBoundedSolidEdge.currentId++;

        this.debugColor = new ColorRgb(1, 1, 1);
    }

    public getEndingVertexId(): number {
        if (this.leftHalf === null) {
            return -1;
        }
        return this.leftHalf.startingVertex.id;
    }

    public getStartingVertexId(): number {
        if (this.rightHalf === null) {
            return -1;
        }
        return this.rightHalf.startingVertex.id;
    }

    public override toString(): string {
        let msg: string;
        msg = "Edge id " + this.id + ". Half1: ";
        if (this.leftHalf === null) {
            msg = msg + "null. ";
        } else {
            msg = msg + "vertex " + this.leftHalf.startingVertex.id;
        }

        msg = msg + " / Half2: ";

        if (this.rightHalf === null) {
            msg = msg + "null. ";
        } else {
            msg = msg + "vertex " + this.rightHalf.startingVertex.id;
        }

        return msg;
    }
}
